third_party/pjs/promise.js

```diff
@@ -0,0 +1,420 @@
+ * reject functions.
+ *
+ * @constructor
+ */
+function Deferred(Promise) {
+    var deferred = this;
+    /** @type {!Promise} */
+    this.promise = new Promise(function(resolve, reject) {
+        /** @type {function(*=)} */
+        deferred.resolve = resolve;
+
+        /** @type {function(*=)} */
+        deferred.reject = reject;
+    });
+    return deferred;
+}
+
+/**
+ * Transitions the state of promise to another state. This is only ever called
+ * on with a promise that is currently in the Pending state.
+ *
+ * @param {!Promise} promise
+ * @param {function(this:Promise,*=,function(*=),function(*=),Deferred):!Promise} state
+ * @param {*=} value
+ */
+function adopt(promise, state, value) {
+    var queue = promise._value;
+    promise._state = state;
+    promise._value = value;
+
+    for (var i = 0; i < queue.length; i++) {
+        var next = queue[i];
+        promise._state(
+            value,
+            next.onFulfilled,
+            next.onRejected,
+            next.deferred
+        );
+    }
+}
+/**
+ * A partial application of adopt.
+ *
+ * @param {!Promise} promise
+ * @param {function(this:Promise,*=,function(*=),function(*=),Deferred):!Promise} state
+ * @returns {function(*=)}
+ */
+function adopter(promise, state) {
+    return function(value) {
+        return adopt(promise, state, value);
+    };
+}
+
+/**
+ * A no-op function to prevent double resolving.
+ */
+function noop() {}
+
+/**
+ * Tests if fn is a Function
+ *
+ * @param {*} fn
+ * @returns {boolean}
+ */
+function isFunction(fn) {
+    return typeof fn === 'function';
+}
+
+/**
+ * Tests if fn is an Object
+ *
+ * @param {*} obj
+ * @returns {boolean}
+ */
+function isObject(obj) {
+    return obj === Object(obj);
+}
+
+/**
+ * Iterates over each element of an array, calling the iterator with the
+ * element and its index.
+ *
+ * @param {!Array} collection
+ * @param {function(*=,number)} iterator
+ */
+function each(collection, iterator) {
+    for (var i = 0; i < collection.length; i++) {
+        iterator(collection[i], i);
+    }
+}
+
+/**
+ * Creates a function that will attempt to resolve the deferred with the return
+ * of fn. If any error is raised, rejects instead.
+ *
+ * @param {!Deferred} deferred
+ * @param {function(*=)} fn
+ * @param {*} arg
+ * @returns {function()}
+ */
+function tryCatchDeferred(deferred, fn, arg) {
+    var promise = deferred.promise;
+    var resolve = deferred.resolve;
+    var reject = deferred.reject;
+    return function() {
+        try {
+            var result = fn(arg);
+            if (resolve === fn || reject === fn) {
+                return;
+            }
+            doResolve(promise, resolve, reject, result, result);
+        } catch (e) {
+            reject(e);
+        }
+    };
+}
+
+/**
+ * Defers fn to another run loop.
+ *
+ * @type {function(function())}
+ */
+var nextTick = typeof setImmediate !== 'undefined' ? setImmediate
```

Comment:
At least in AMP it is easy to do a polyfill based on postMessage to our own window.

Basically do

`window.addEventListener('message', flush)` and `window.postMessage('macro-task')`.

This is only relevant for older iOS and Android, but we want decent support there.